tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,387 @@
+Disclaimer: this script attempts to be as faithful as possible to the
+`chpldoc.cpp` implementation, but may differ slightly in behavior.
+
+The script works on the following symbols
+- Modules
+- Functions
+  - In this sense, a function is anything declared with `proc`, `iter`, or
+    `operator`
+  - This includes functions defined inside of Modules, AggregateDecls, and
+    Interfaces
+  - This will not match nested Functions
+- AggregateDecls
+  - AggregateDecls are Records, Classes, or Unions
+  - This will match nested definitions in Modules and AggregateDecls,
+    e.g. a Record inside a Record
+- Decls
+  - Decls are top level variables defined in Modules or AggregateDecls
+- MultiDecls
+  - Same as Decls, but multiple symbols share the same keyword, e.g.
+    `var a, b = 10;`
+- Enums
+  - Will also match against EnumElements
+  - This will match nested definitions in Modules and AggregateDecls
+- Interfaces
+  - This will match nested definitions in Modules and AggregateDecls
+
+The script pulls out all of these symbols from a parsed source file, and then
+determines if the symbol is documented. A symbol is considered documented if
+the previous sibling node in the AST is a docstring Comment.
+
+Symbols can be marked 'nodoc', meaning it is ok for the docstring to be
+missing. This script considers a symbol to be marked 'nodoc' based on the
+following criteria:
+- If it is explicitly marked with an attribute `@chpldoc.nodoc`
+- If its names is prefixed with `chpl_`
+  - This also checks the `pragma "chpldoc ignore chpl prefix"`
+- If the symbol is marked private
+- If the symbol has a parent that is marked 'nodoc'
+
+This script also provides two optional flags, `--ignore-[unstable|deprecated]`.
+These work based on explicit `@unstable` or `@deprecated` attributes, if the
+symbol was marked in a different way this script will still warn. If a symbol's
+parent was marked unstable or deprecated then the symbol is also considered to
+be unstable or deprecated.
+"""
+
+from collections import defaultdict
+from typing import Generator, List
+import sys
+import os
+import argparse as ap
+import itertools
+
+try:
+    import chapel.core as dyno
+    from chapel import each_matching
+except ModuleNotFoundError:
+    print("chapel python bindings not found")
+    exit(1)
+
+
+"""
+helpers
+"""
+
+
+def get_module(n: dyno.AstNode) -> dyno.AstNode:
+    if n is None:
+        return None
+    if isinstance(n, dyno.Module):
+        return n
+    if parent := n.parent():
+        return get_module(parent)
+    return None
+
+
+def node_or_parent_has_attribute(node: dyno.AstNode, marker: str) -> bool:
+    """
+    a symbol is has an attribute if it has an attribute that matches the argument, or is parent does
+    """
+    if (attrs := node.attribute_group()) and any(a.name() == marker for a in attrs):
+        return True
+
+    if (parent := node.parent()) and node_or_parent_has_attribute(parent, marker):
+        return True
+
+    return False
```

Comment:
Should we warn for:

```Chapel
@deprecated
module M {
  module N {
    proc noComment() {

    }
  }
}
```

My intuition says no, but checking only the direct parent seems like it would cause a warning here.

A more principled approach (and one that perhaps won't require looking up parents and could save some work) is to simply not recurse into nodes that are nodoc'ed in one way or another. This would require a modified version of `each_matching` that fuses "find documentable symbols" with "report undocumented symbol", but neither of those are that fancy (I think).

Reply:
The current approach will not warn for `M`, `N`, or `noComment` if `--ignore-deprecated` is used. I think this is the behavior we want, but I take your point on the more principled approach.

Reply:
I implemented a better version where `each_matching` can now take an `iterator` optional argument to define what nodes it traverses